### PR TITLE
Deprecation Work - Replace JException

### PR DESCRIPTION
**Pull request by mbabker**
This pull request removes most instances of JException in favor of Exception.  All but the Form package are handled because I couldn't sort out where the unit test suite was failing with this change in that package.

### CODE REVIEW

File: libraries/joomla/database/table.php [hunk 1]
```diff
@@ -118,6 +118,9 @@ public function __construct($table, $key, &$db)
 	 * Get the columns from database table.
 	 *
 	 * @return  mixed  An array of the field names, or false if an error occurs.
+	 *
+	 * @since   11.1
+	 * @throws  Exception
```
Comment (realityking) [inline libraries/joomla/database/table.php]: The exception is never actually throw is it? It's just returned.

**Comment by realityking** [comment]
Have you tested this with the CMS? JException seems to offer some methods Exception doesn't, I wonder if they are actually used.

**Comment by mbabker** [comment]
Looked through the CMS.  In instances where JException is used, it only uses the first two params, which are the same as Exception.  There may be an instance or two where the catch statement needs to be changed, otherwise, there should be no issues.

File: libraries/joomla/database/table.php (hunk 1 above)
Comment (eddieajau) [inline libraries/joomla/database/table.php]: Aye. Should remove the @throws.

**Comment by LouisLandry** [comment]
This is fantastic.  Can you give me an idea of what you were running into problems with in the form package?  I may be able to help.

**Comment by mbabker** [comment]
The `$valid->get('level')` switch at https://github.com/joomla/joomla-platform/blob/staging/libraries/joomla/form/form.php#L1119

This might work out correctly now that #604 has been merged.

**Comment by LouisLandry** [comment]
Ah.

That probably needs to be rewritten to look more like:

``` <?php
    // Attempt to validate the field.
    try
    {
        $this->validateField($field, $group, $value, $input);
    }
    // There was an invalid field due to unexpected value, etc.
    catch (RuntimeException $e)
    {
        array_push($this->errors, $e);
        $return = false;
    }
    // There was a problem actually validating the field.
    catch (LogicException $e)
    {
        JError::raiseWarning(0, $e->getMessage());
        return false;
    }
```

What this would require is that form fields actually throw exceptions instead of returning them.  I don't know that we can just do that right now, but we can certainly mark some stuff deprecated and start handling that case.

**Comment by realityking** [comment]
@mbabker
It should have helped. The code I substituted is 1:1 what is in JError::isError()

**Comment by realityking** [comment]
How about this for 11.4:

```
// Attempt to validate the field.
try
{
    $valid = $this->validateField($field, $group, $value, $input);

    // @deprecated 12.1
    if ($valid instanceof JException)
    {
        JLog::add('JFormRule subclasses should not return JExceptions. Instead either a RuntimeException or LogicException should be thrown. ', JLog::WARNING, 'deprecated');
        switch ($valid->get('level'))
        {
            case E_ERROR:
                JError::raiseWarning(0, $valid->getMessage());
                return false;
                break;

            default:
                array_push($this->errors, $valid);
                $return = false;
                break;
        }
    }
}
// There was an invalid field due to unexpected value, etc.
catch (RuntimeException $e)
{
    array_push($this->errors, $e);
    $return = false;
}
// There was a problem actually validating the field.
catch (LogicException $e)
{
    JFactory::getApplication()->enqueueMessage($e->getMessage(), 'warning');
    return false;
}
```

**Pull request closed without merging**